sjforman/votefwd
# Set up Ohio experiment and targeting

Hey @andrewjtimmons, I think you might be 😷 according to Instagram, so no worries if you're not up to at the moment, but if you are, would you take a look at this and see if it makes sense / if there's anything you think I should change or add? 

I took a meandering path here, initially intending to add a notion of a "variant" and to have multiple variants per experiment. But that got complicated quickly, and I wound up concluding not worth the effort right now. So, key things this does add: 

 * A migration to make a table to store raw Ohio data in a table called `ohio_12_wide`. Raw data from Ohio_12 can be loaded from postgres command line with: 

```sql    
COPY ohio_12_wide from '[PATH to csv on your computer]' DELIMITER ',' CSV HEADER;
```

 * A migration that adds tables called `experiment` and `experiment_voter`. 

 * A sql script that can be run from the command line that populates `experiment` and `experiment_voter` with data from `ohio_12_wide`, and then, in turn, populates the `voter` table with some randomly selected folks from `experiment_voter`. 

* Adds "suffix" to `voter` table and displays it in the appropriate places. 

Note that I'm not explicitly designating anyone as a `CONTROL` for this experiment, but I'm refraining on purpose. Because the total addressable population is ~30k, and I'm anticipating that we won't get close to that in this experiment, I intend to pull additional voters as needed into the 'TEST' group as needed. I'm thinking I'll designate everyone left at the end as `CONTROL`. 

This is all a stopgap as discussed, because not only does this not handle data from other districts in Ohio, it won't even handle the updated file we'll get from Ohio after the election. It also occurs to me that I don't know exactly how I will run that sql script against a deployed environment, and that it doesn't yet handle ad-hoc additions to the ranks of the `TEST` group. Can figure all that out later, though. 

How to test: 
 * Run migrations. 
 * Load Ohio data `COPY ohio_12_wide from '[PATH to csv on your computer]' DELIMITER ',' CSV HEADER;`
 * Run targeting script. `psql -f ../bin/ohio-targeting.sql votefwd`
 * `select(*) from experiment;` and confirm it looks as intended. 
 * `select(*) from experiment_voter;` and confirm it looks as intended. 
 * `select(*) from voters;` and confirm it looks as intended.
 * Start app. 
 * Login. 
 * Adopt a voter. 
 * See that it's one of the Ohio voters and that details match the row from the original CSV.
